### PR TITLE
Introduce abstract base classes for store operations and nodes

### PR DESCRIPTION
**Pull request by phate**
The PR does the following:
1. Introduces an abstract base class for store operations
2. Introduces an abstract base class for store nodes
3. Factors out common functionality for the store classes

### CODE REVIEW

File: jlm/llvm/ir/operators/Store.cpp [hunk 1]
```diff
@@ -63,6 +97,41 @@ StoreVolatileOperation::copy() const
...
+
+  return { MemoryStateInputIterator(input(3)), MemoryStateInputIterator(nullptr) };
+}
+
+[[nodiscard]] StoreNode::MemoryStateOutputRange
+StoreVolatileNode::MemoryStateOutputs() const noexcept
+{
+  if (NumMemoryStates())
+  {
+    return { MemoryStateOutputIterator(nullptr), MemoryStateOutputIterator(nullptr) };
```
Comment (haved) [inline jlm/llvm/ir/operators/Store.cpp:129]: This is backwards, no? If there are memory states, the iterator pair should not start with nullptr?

File: jlm/llvm/ir/operators/Store.cpp [hunk 2]
```diff
@@ -33,6 +33,40 @@ StoreNonVolatileOperation::copy() const
...
+
+  return { MemoryStateInputIterator(input(2)), MemoryStateInputIterator(nullptr) };
+}
+
+[[nodiscard]] StoreNode::MemoryStateOutputRange
+StoreNonVolatileNode::MemoryStateOutputs() const noexcept
+{
+  if (NumMemoryStates())
+  {
+    return { MemoryStateOutputIterator(nullptr), MemoryStateOutputIterator(nullptr) };
```
Comment (haved) [inline jlm/llvm/ir/operators/Store.cpp:64]: This is backwards, no? If there are memory states, the iterator pair should not start with nullptr?

**Review by haved — CHANGES_REQUESTED** [review]
I think there might be to if statements with inverted conditions

File: jlm/llvm/ir/operators/Store.cpp (hunk 1 above)
Comment (phate) [reply jlm/llvm/ir/operators/Store.cpp:129]: Really great catch!

File: jlm/llvm/ir/operators/Store.cpp (hunk 2 above)
Comment (phate) [reply jlm/llvm/ir/operators/Store.cpp:64]: Well spotted!

**Comment by phate** [comment]
@haved I need another approval. I had a merge conflict.

**Review by haved — APPROVED** [review]
Looks good!

**Pull request merged**